src/v/cloud_storage/partition_manifest.cc

```diff
@@ -856,15 +856,29 @@ void partition_manifest::reset_scrubbing_metadata() {
 
 std::optional<size_t> partition_manifest::move_aligned_offset_range(
   const segment_meta& replacing_segment) {
+    // nullopt means that the replacing_segment can't be inserted in the
+    // manifest, either by breaking an invariant in segment_map, or by failing
+    // to generate a distinct name
+    if (unlikely(!_segments.can_be_inserted(replacing_segment))) {
+        // log an error, as it's not expected for a segment at this stage to
+        // fail this check. safe_segment_meta_to_add() fail before this
+        vlog(
+          cst_log.error,
+          "{} segment {:s} can't be inserted because it's covered by another "
+          "in the manifest",
+          _ntp,
+          replacing_segment);
+        return std::nullopt;
+    }
+
     size_t total_replaced_size = 0;
+
     auto replacing_path = generate_remote_segment_name(replacing_segment);
```

Comment:
I think I follow the previous commit, but I'm having a hard time determining whether this is a functional change or not. Can you clarify?

Reply:
so the intent is to _not_ introduce a functional change (the whole pr tries to make sure to not touch the correct behavior, just to restrict a bad behavior from happening).

with can_be_inserted, i am able to determine another case where move_aligned_offset_range should return nullopt. The whole redpanda subsystem should prevent this case from happening at all, so i see this block as code documenting explicitly an implicit behavior.

with covered_range, the change stems from the comment made by vlad that noticed how the logic to determine the replaced segments is duplicated in partition_manifest and segment_meta cstore, so i created covered_range to keep the code in a single translation unit. 

I can probably split this commit in two